File: Presentations/NonNormalData_hfe/demo_app/Bootstrap-sim-ex.java
// Java program for visualization bootstrap calculations using simulations
//   by: <harald.weedon-fekjar@medisin.uio.no> 2022
//   Java : https://en.wikipedia.org/wiki/Java_(programming_language)

import java.util.Scanner;
import java.util.Random;
import java.util.Arrays;
import java.util.ArrayList;

class BootstrapSimEx {

  private static float[] introAskData() {
    System.out.print("\n\n\n");
    System.out.print("Welcome to an interative demostration of statistical bootstrapping");
    System.out.print("\n\n");
    Scanner input = new Scanner(System.in);
    System.out.print("Input comma seperated data: ");
    String datat = input.nextLine();
    System.out.println(datat);
    String[] datats = datat.split(",");
    float[] data = new float[datats.length];
    for (int i = 0; i < datats.length; ++i) {
      data[i] = Float.parseFloat(datats[i]);
    }
    return data;
  }

  private static float OneBootsample(Random rn, float[] dataA, boolean printres, int nb) {
    int noobs = dataA.length;
    float bootSum = 0;
    ArrayList<Float> boots = new ArrayList<Float>();
    String bootstr = "";
    for (int i = 0; i < noobs; i++) {
      int indno = rn.nextInt(noobs);
      float newval = dataA[indno];
      boots.add(newval);
      bootSum += newval;
      bootstr += String.valueOf(newval);
      if (i < noobs - 1) {
        bootstr += ", ";
      }
    }
    float bootsMean = bootSum / noobs;
    if (printres == true) {
      System.out.println("Bootstrap sample " + nb + ":");
      System.out.println(" Bootstrap sample: " + bootstr);
      System.out.printf(" Mean: %.1f\n", bootsMean);
    }
    return bootsMean;
  }

  public static void main(String[] args) {
    int noboot = 10000;
    float res[] = new float[noboot];
    boolean printRes = true;
    Random rn = new Random(3);
    Scanner input = new Scanner(System.in);
    float[] data = introAskData();
    System.out.println("\n\n\nRuning bootstrap ...");
    System.out.println("Hit <return> for new sample, S+<return> for sumerize/end\n");
    for (int i = 0; i < noboot; i++) {
      res[i] = OneBootsample(rn, data, printRes, i + 1);
      if (printRes) {
        String ans = input.nextLine();
        if (ans.length() > 0) {
          if (ans.charAt(0) == 'S' || ans.charAt(0) == 's') {
            printRes = false;
          }
        }
      }
    }
    Arrays.sort(res);
    float ciP1 = noboot *(float)0.025 - (float)1;
    float ci1 = res[(int) ciP1];
    float ciP2 = noboot *(float)0.975 - (float)1;
    float ci2 = res[(int) ciP2];
    System.out.println("...\n...\n...\n...\n...\n\n");
    System.out.print("\n\n95% bootstrap percentile confidence interval: ");
    System.out.printf("[%.1f, %.1f]\n",ci1,ci2);
    System.out.printf("(based on %d bootstrap replications)\n\n", noboot);
    String finished = input.nextLine();
  }
}

File: Presentations/NonNormalData_hfe/demo_app/Bootstrap-sim-ex.cs
// C# program for visualization bootstrap calculations using simulations
//   by: <harald.weedon-fekjar@medisin.uio.no> 2022
//   C# : https://en.wikipedia.org/wiki/C_Sharp_(programming_language)

  //// Principle:
  // - Ask user for comma separated data
  // - Shows bootstrap sample with observations and mean
  //   User asked to answer:
  //   N = Next sample, S = Show summarize confidence intervals

  ////  Run and build with:
  // dotnet run
  // dotnet build  // dotnet publish -r win-x64

using System;
using System.Collections.Generic;

class Program {
  
  static decimal OneBootsample(Random rand,decimal[] dataA, bool printRes,int nb) {
    int noobs = dataA.Length;
    var boots = new List<decimal>();
    for (int i=0; i<noobs; i++) {
      int indno = rand.Next(0,noobs);
			boots.Add(dataA[indno]);
		}
    decimal bootsMean = boots.Average();
    if (printRes) {
      Console.WriteLine("");
      Console.WriteLine("Bootstrap sample " + nb + ":");
      Console.WriteLine("   Bootstrap sample: "+String.Join(",",boots));
      Console.WriteLine("   Mean:             " + Math.Round(bootsMean,2));
    }
		return bootsMean;
  }

  public static void Main (string[] args) {
    int noboot = 10000;
    // Leser inn data:
    Console.Write("\n\n\n");
    Console.Write("Welcome to an interative demostration of statistical bootstrapping");
    Console.Write("\n\n");
    Console.Write("Input comma seperated data: ");
    string dataText = Console.ReadLine()!;
    decimal[] dataA = dataText.Split(',').Select(decimal.Parse).ToArray();
    // Kaller på bootstrap rutine:
    Console.WriteLine("\n\n\nRuning bootstrap ...");
    Console.WriteLine("Hit <return> for new sample, S for sumerize/end\n");
    var rand = new Random(3); 
    decimal  newres;
    char printA='n';
    bool printRes=true;
    var Res = new List<decimal>();
    for (int i=0; i<noboot; i++) {
      newres  = OneBootsample(rand,dataA,printRes,i+1);
      Res.Add(newres);
      if (printRes==true) { 
        printA = Console.ReadKey().KeyChar; 
      }
      if (printA=='S' || printA=='s') { printRes=false; }
    }
    // Oppsummerer:
    Res.Sort();
    decimal ci1 = Res[Convert.ToInt32(noboot*0.025-1)];
    decimal ci2 = Res[Convert.ToInt32(noboot*0.975-1)];
    Console.Write("...\n...\n...\n...\n...\n\n");
    Console.ForegroundColor = ConsoleColor.Blue;
    Console.Write("\n\n95% bootstrap percentile confidence interval: ");
    Console.WriteLine("[{0}, {1}]",Math.Round(ci1,2),Math.Round(ci2,2));
    Console.ResetColor();
    Console.Write("(based on {0} bootstrap replications)\n\n",noboot);
    char finishYes;
    finishYes = printA = Console.ReadKey().KeyChar;
  }
}
